### PR TITLE
Add examples building as jupyter notebooks into the documentation

### PR DESCRIPTION
**Pull request by kudep**
rework of #20

### CODE REVIEW

**Comment by pseusys** [comment]
It was merged already, wasn't it?

**Comment by kudep** [comment]
Nope

**Comment by RLKRo** [comment]
Merge was reverted due to several issues such as:

- [ ] pr#20 didn't have a proper code review
- [ ] pr#20 partially reverted pr#18 (it increased pipeline testing times)

File: makefile [hunk 1]
```diff
@@ -87,15 +87,17 @@ version_major: venv
 	bump2version --current-version $(CURRENT_VERSION) major $(VERSIONING_FILES)
 .PHONY: version_major
 
+clean_docs:
+	rm -rf docs/build
+	rm -rf docs/examples
+	rm -rf docs/source/apiref
+	rm -rf docs/source/examples
+.PHONY: clean
```
Comment (RLKRo) [inline makefile:95]: Should be `.PHONY: clean_docs`.

File: tests/pipeline/test_examples.py [hunk 2]
```diff
@@ -1,10 +1,14 @@
 import importlib
+import os
+import sys
 
 import pytest
 
 import tests.utils as utils
 
-dot_path_to_addon = utils.get_dot_path_from_tests_to_current_dir(__file__)
+sys.path.append(os.path.abspath(f"examples/{utils.get_path_from_tests_to_current_dir(__file__)}"))
```
Comment (RLKRo) [inline tests/pipeline/test_examples.py:9]: Why is this line added? It is not necessary since all examples are imported with an `examples.{dot_path_to_addon}` prefix.
Also it won't work on Windows.

**Review by RLKRo — CHANGES_REQUESTED** [review]
Some of the comments I added apply to multiple files (e.g. my comment on [examples/pipeline/1_basic_example.py](https://github.com/deeppavlov/dialog_flow_framework/pull/21/files/40f56b6a91cd9b20b83e93343f1946d52a0ff48c#diff-64e3e7a3b750f42b879a9f129628aa9e3ebe7ad3ca641787b83203e54490ed1d) is also relevant for other pipeline examples).

File: docs/source/conf.py [hunk 3]
```diff
@@ -53,10 +47,11 @@
     "sphinx.ext.viewcode",
     "sphinx.ext.mathjax",
     "sphinx.ext.extlinks",
-    #    "sphinxcontrib.katex",  # TODO: throws an exception for some reason
+    "sphinxcontrib.katex",
```
[the remark this answers is not in the document]
Comment (pseusys) [reply docs/source/conf.py:50]: Something like "file /temp/temp-[some_hash].txt not found". I think there were some internal conflicts between katex and sphinx_gallery. Anyway, it's not here anymore.

File: docs/source/generate_notebook_links.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            elif entity.is_dir() and not entity.name.startswith("_"):
+                if len(process_dir(entity, exclude)) > 0:
+                    includes.append(f"{doc_path.name}/index")
+    if len(includes) > 0:
+        create_index(Path("docs/source/examples") / path.relative_to("examples/") / Path("index.rst"), includes)
+    return includes
+
+
+def generate_links(include: Optional[List[str]] = None):
+    process_dir(Path("examples/"), ["**"] if include is None else include)
```
[the remark this answers is not in the document]
Comment (pseusys) [reply docs/source/generate_notebook_links.py:54]: Some merge conflicts

File: examples/pipeline/1_basic_example.py [hunk 5]
```diff
@@ -1,20 +1,21 @@
...
 """
 
+# %%
 import logging
 
 from dff.core.engine.core import Context
 
 from dff.core.pipeline import Pipeline
-from examples.pipeline._pipeline_utils import SCRIPT, get_auto_arg, auto_run_pipeline
+from _pipeline_utils import SCRIPT, should_auto_execute, auto_run_pipeline
```
[the remark this answers is not in the document]
Comment (pseusys) [reply examples/pipeline/1_basic_example.py:13]: It is required for documentation building.
In a word: during documentation building we create soft links under `docs/sources` and execute example files there. There is no `example.pipeline` module in there, so we face an import error.

File: examples/pipeline/5_asynchronous_groups_and_services_basic.py [hunk 6]
```diff
@@ -38,7 +38,7 @@
 
 
 async def time_consuming_service(_):
-    await asyncio.sleep(0.01)
+    await asyncio.sleep(1)
```
[the remark this answers is not in the document]
Comment (pseusys) [reply examples/pipeline/5_asynchronous_groups_and_services_basic.py]: Some merging conflicts again, wasn't done intentionally

File: makefile (hunk 1 above)
Comment (pseusys) [reply makefile:95]: 👍

File: tests/pipeline/test_examples.py (hunk 2 above)
Comment (pseusys) [reply tests/pipeline/test_examples.py:9]: Please, check workflow runs - it works on Windows.
It _is_ necessary, since there are relative imports in the imported examples.

**Review by RLKRo — APPROVED** [review]

File: tests/pipeline/test_examples.py (hunk 2 above)
Comment (kudep) [reply tests/pipeline/test_examples.py:9]: This is necessary because of  `from _pipeline_utils import SCRIPT, should_auto_execute, auto_run_pipeline
`

Comment (kudep) [reply tests/pipeline/test_examples.py:9]: Pls, add `TODO` to remove when we will add example utils into pipy package
